### PR TITLE
[JENKINS-40803] provide a config file resolution context for promotions

### PR DESCRIPTION
**Pull request by Greybird**
downstream proposal for jenkinsci/config-file-provider-plugin#49.
to implement solution proposed for #104

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -123,6 +129,12 @@
       <version>1.4.7-jenkins-1</version>
       <optional>true</optional>
     </dependency>
+    <dependency>
+      <groupId>org.jenkins-ci.plugins</groupId>
+      <artifactId>config-file-provider</artifactId>
+      <version>2.17.1-SNAPSHOT</version>
```
Comment (oleg-nenashev) [inline pom.xml]: You could use "maven deploy" to deploy a timestamped snapshot for testing this PR in CI

File: pom.xml [hunk 2]
```diff
@@ -75,6 +75,12 @@
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>junit</artifactId>
       <version>1.20</version>
+      <exclusions>
+        <exclusion>
+          <groupId>org.jenkins-ci.plugins</groupId>
+          <artifactId>structs</artifactId>
```
Comment (oleg-nenashev) [inline pom.xml]: bad idea, it's better to add explicit dependency on Structs plugin to this POM

File: src/main/java/hudson/plugins/promoted_builds/JobPropertyImplConfigContextResolver.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package hudson.plugins.promoted_builds;
+
+import hudson.Extension;
+import hudson.model.ItemGroup;
+import org.jenkinsci.plugins.configfiles.ConfigContextResolver;
+
+@Extension(optional = true)
+public class JobPropertyImplConfigContextResolver extends ConfigContextResolver {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/promoted_builds/JobPropertyImplConfigContextResolver.java]: Add `@Restricted(NoExternalUse.class)`

**Comment by Greybird** [comment]
Thanks for the review @oleg-nenashev 
Did the requested changes, can you review again?

**Comment by Greybird** [comment]
Following release of 2.18 version of config-files-provider plugin, I updated the PR to use this version.
Thanks for your help.

**Comment by dwnusbaum** [comment]
@Greybird Thanks! I am planing on merging and releasing the change tonight.

**Pull request merged**